client/lib/logmein/index.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * Internal Dependencies
+ */
+import { isEnabled } from '@automattic/calypso-config';
+import wpcom from 'calypso/lib/wp';
+
+/**
+ * Append logmein=1 query parameter to mapped domain urls we want the user to be logged in against.
+ */
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
+	let newurl: URL;
+
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	let allow = logmeinAllowedUrls( sites );
+	// Ignore urls not in the allow list
+	allow = allow.map( ( allowed: string ) => new URL( allowed, INVALID_URL ).hostname );
+	if ( allow.indexOf( newurl.hostname ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// we're already logged into wordpress.com so we can take a shortcut
+	if ( isWPComLoggedIn ) {
+		newurl.searchParams.set( 'logmein', 'direct' );
+	} else {
+		newurl.searchParams.set( 'logmein', '1' );
+	}
+	return newurl.toString();
+}
+
+export function logmeinAllowedUrls( sites: any ): string[] {
+	return sites
+		.map( ( site: any ) =>
+			! site.is_vip &&
+			! site.jetpack &&
+			! site.options.is_automated_transfer &&
+			! site.options.is_domain_only &&
+			! site.options.is_redirect &&
+			! site.options.is_wpcom_atomic &&
+			! site.options.is_wpcom_store &&
+			! site.options.is_wpforteams_site &&
+			site.options.is_mapped_domain
+				? new URL( site.URL ).host
+				: false
+		)
+		.filter( Boolean );
+}
+
+export function appendLogmein( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', '1' );
+	return url;
+}
+export function appendLogmeinDirect( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', 'direct' );
+	return url;
+}
+
+let allowedSites: string[] = [];
```

Comment:
Is there a better place/way to store this response for use by the navigate wrapper and onclicker?

Reply:
You can replace the `wpcom.me().sites()` request by reading from Redux store and relying on someone else issuing the request for you. We currently always issue the request in the main layout except some Jetpack Connect routes.

Your click handler can do something like:
```js
function onClick( event ) {
  const link = event.target.href;
  const sites = Object.values( getSitesItems( reduxStore.getState() ) );
  if ( sites.some( site => link.startsWith( site.URL ) ) ) {
    event.target.href = addLogmeinQuery( link );
  }
}
```
That will use the latest known sites list on each click.

Reply:
In `client/layout/index.jsx` we render a `<QuerySites allSites />` component which does the request on mount. And there are several places that dispatch the `requestSites` Redux action to refresh the sites list, e.g., after accepting an invite to a new one etc.

Reply:
Will we need to attach `logmein` query params in situations where the Calypso redux store doesn't exist? For example the ETK plugin or the notifications popup iframe?

If so, then what about if `attachLogmein()` accepted an `allowedSites` parameter? If it's missing then `attachLogmein()` will have to make the request itself. If it's provided then it doesn't have to make the extra request.

When `attachLogmein()` is called in `client/boot/common`, since it has access to the redux store right there, it can pass in the allowed sites.

Reply:
But since this is in `client/lib/logmein`, perhaps it's only meant to be used inside of Calypso.

Reply:
Will we need to attach logmein query params in situations where the Calypso redux store doesn't exist?

[Yes, see editor iframe link comments re ETK.](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647925730)

Other situations less likely but I would like the option to leverage at least some of the same code.

Reply:
Tackling this in a follow up.